src/compiler/checker.ts

```diff
@@ -4716,19 +4731,33 @@ namespace ts {
             return result;
         }
 
+        function cloneTypePredicate(predicate: TypePredicate, mapper: TypeMapper): ThisTypePredicate | IdentifierTypePredicate {
+            if (predicate.kind === TypePredicateKind.Identifier) {

```

Comment:
I would really prefer we didn't use a magic number because you'd be special-casing in a slightly different but more implicit way.